src/FGraphDiff/mrob/factor_diff.hpp

```diff
@@ -0,0 +1,99 @@
+/* Copyright (c) 2022, Gonzalo Ferrer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ * factor.hpp
+ *
+ *  Created on: Feb 12, 2018
+ *      Author: Gonzalo Ferrer
+ *              g.ferrer@skoltech.ru
+ *              Mobile Robotics Lab, Skoltech 
+ */
+
+#ifndef FACTOR_DIFF_HPP_
+#define FACTOR_DIFF_HPP_
+
+#include <vector>
+#include <memory>
+
+#include "../FGraph/mrob/factor.hpp"
+
+
+namespace mrob{
+
+/**
+ * DiffFactor class is a base pure abstract class defining factors,
+ * the second type of vertexes on factor graphs (bipartite).
+ * Factors keep track of all their neighbour nodes they are connected to.
+ *
+ * By convention, the residuals r_i are ALWAYS formulated as follows:
+ *
+ * -------------------------------------------
+ * |           r(x) =  h(x) - z              |
+ * -------------------------------------------
+ *
+ * otherwise the optimization will not work properly.
+ */
+
+
+class DiffFactor : public Factor
```

Comment:
I think this is the main issue. You need to get jacobians, so it is not enough to evaluate, you need to create methods to pass the partial derivatives.